pyro/distributions/torch.py

```diff
@@ -282,11 +282,28 @@ def expand(self, batch_shape):
             return type(self)(df, loc, scale, validate_args=validate_args)
 
 
-class TransformedDistribution(torch.distributions.TransformedDistribution, TorchDistributionMixin):
+class TransformedDistribution(torch.distributions.TransformedDistribution, TorchDistributionMixin, torch.nn.Module):
+    def __init__(self, *args, **kwargs):
+        torch.distributions.TransformedDistribution.__init__(self, *args, **kwargs)
+        TorchDistributionMixin.__init__(self)
+        torch.nn.Module.__init__(self)
+
+        # Not all transforms have parameters, so only want to extract nn.Module ones
+        self.transform_modules = torch.nn.ModuleList([t for t in self.transforms if isinstance(t, torch.nn.Module)])
```

Comment:
This will register the parameters of the transform for use in `.parameters` and `.cuda`. Are there cases where we would have to separately grab the base distribution's parameters? If so, it might give a false impression to the user that they have transferred their parameters to the GPU by calling `transformed_dist.cuda()`, but they'll get an error later because their base distribution's params will still be on the CPU.

Reply:
Hmm, at the moment you can't do `.cuda()` or `.parameters()` on a distribution (Fritz explains the rationale [here](https://github.com/pytorch/pytorch/issues/7795#issuecomment-391807661)). I guess the difficulty is created because a transformed distribution mixes objects with and without trainable parameters... I've added in an extra bit of code that adds a base distribution that is also an `nn.Module` to the module list

Reply:
I agree it would be confusing if a user calls `.cuda()` and their base distribution in on the `.cpu()`. Although I can't see a good solution. It doesn't seem possible to add a `.cuda()` to distributions in general for the reasons in that issue above. On the other hand, there's a need to have `.cuda()` for normalizing flows, since they have trainable parameters... What about overriding `.cuda()`, `.to()` and `.cpu()` in `TransformedDistribution` to raise an exception? Then the preferred way to transform your normalizing flow to the GPU would be to call `.cuda()` on each `TransformModule` separately (or in a list comprehension)?

Reply:
The more I think about this, the more I prefer not making `TransformedDistribution` an `nn.Module`, keeping it lightweight like `Distribution`, and doing all the work with `TransformModule`. I was just looking over the DMM example in Pyro, and I think in a typical use case, you would make an `nn.ModuleList` of `Transforms` inside your model (inheriting from `nn.Module`), and create a `TransformedDistribution` on the fly as needed

Reply:
I agree - I think that will be a cleaner, more straightforward abstraction!